frame/treasury/src/types.rs

```diff
@@ -0,0 +1,116 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::{fmt::Display, marker::PhantomData};
+use sp_std::fmt::Formatter;
+
+use codec::{Decode, Encode, FullCodec, MaxEncodedLen};
+use frame_support::{traits::tokens::Balance, RuntimeDebug};
+use scale_info::TypeInfo;
+
+// At the moment when using PartialEq on AssetId, native
+// is expected to be loweset.
+pub trait MultiAssetBalanceConverter {
+	type AssetId;
+	type NativeBalance;
+	type AssetBalance;
+	type NativeOrAssetBalance;
+
+	fn get_native(balance: Self::NativeBalance) -> Self::NativeOrAssetBalance;
+
+	fn try_convert_asset(
+		asset: Self::NativeOrAssetBalance,
+	) -> Result<(Self::AssetId, Self::AssetBalance), ()>;
+
+	fn try_convert_native(asset: Self::NativeOrAssetBalance) -> Result<Self::NativeBalance, ()>;
+
+	fn into_multiasset_balance(
+		asset: Self::AssetId,
+		balance: Self::AssetBalance,
+	) -> Self::NativeOrAssetBalance;
+}
+
+/// An implementation of MultiAssetId that chooses between Native and an asset.
+#[derive(
+	Decode,
+	Encode,
+	PartialEq,
+	Eq,
+	MaxEncodedLen,
+	TypeInfo,
+	Clone,
+	Copy,
+	Ord,
+	PartialOrd,
+	RuntimeDebug,
+)]
+pub enum NativeOrAssetBalance<AssetId, NativeBalance, AssetBalance> {
```

Comment:
can we stop doing this? it shouldn't be the treasury's job to handle the difference between native and other assets. just build another pallet/helper for it.

Reply:
I'm curious though. How would you approach it in a practical sense. Assuming it's a new pallet, how would it work?

Reply:
https://github.com/paritytech/substrate/pull/12984#issuecomment-1362139263

Reply:
Thanks! I've gone through it, and I'm exploring introducing a separate pallet for the multi asset functionality, which in turn actually makes things easier